### PR TITLE
[RFC] ST sensor drivers migration to STdC

### PR DESCRIPTION
**Pull request by avisconti**
STdC is a standard package written by ST that aims to provide a standard register i/f for ST MEMS
sensors. It is already offered/used in other s/w systems as well.
The idea behind it is to have a clean and stable functions/defines to read/write sensors registers.
Official public repo is: 
https://github.com/STMicroelectronics/STMems_Standard_C_drivers

The license seems to be the same as stm32cube, so there should be no issues here.
Since STdC package is including math.h I had to enable NEWLIB_LIBC.

In this RFC I did an experimental test using lis2dw12 driver. It works for both i2c and spi.
Please provide your feedback.

### CODE REVIEW

**Comment by avisconti** [comment]
@erwango 
This is not urgent as I'm planning it for v1.15, but would be great to receive your feedback
for the 'ext/' part as soon as possible. If it makes sense to proceed with this standard sensor reg i/f 
drivers I have some porting work to do.

**Comment by avisconti** [comment]
@MaureenHelm 
I would like to seek your opinion as well.

**Comment by avisconti** [comment]
@erwango @MaureenHelm 
I re-aligned to latest commit and re-pushed.
As I was saying this RFC is quite important for me, as if accepted I would need to make modifications
to #13062 and to #13478 outstanding PR.

**Review by erwango — CHANGES_REQUESTED** [review]
Few comments on packages integration, otherwise looks good.

File: ext/hal/st/stmemsc/README
Comment (avisconti) [reply ext/hal/st/stmemsc/README]: We have the same agreement cited in all stm32cube packages, like ext/hal/st/stm32cube/stm32f4xx/README for example.

Why should be an issue here?

File: ext/hal/st/stmemsc/README.md
Comment (avisconti) [reply ext/hal/st/stmemsc/README.md]: Correct! I removed all the examples. So I will strip off from the README.md all references to examples.

File: ext/hal/st/stmemsc/a3g4250d_STdC/README.txt
Comment (avisconti) [reply ext/hal/st/stmemsc/a3g4250d_STdC/README.txt]: Mmmmh ok.
So I see two options now. Either keeping also the examples in the package or adapting all the
README (but there are many) stripping of all references to examples.

@erwango 
If you don't see issues I may go with option one which seems to be less impacting.

File: ext/hal/st/stmemsc/README
Comment (erwango) [reply ext/hal/st/stmemsc/README]: Right, let me sort this out.

File: ext/hal/st/stmemsc/a3g4250d_STdC/README.txt
Comment (erwango) [reply ext/hal/st/stmemsc/a3g4250d_STdC/README.txt]: I'm not sure adding examples juts because README files refer to it is a great option.
Do we really need these README files at all?
I don't see much added value in the information provided, this is only copy/paste from one to the other, and most of the information is not valid as referring to examples that you initially considered as superfluous (and I think you were right).
The only interesting information that differ from one file to the other is `H3LIS331DL DS rev3.0 `, which could be factorized in a upper file if needed.

Comment (avisconti) [reply ext/hal/st/stmemsc/a3g4250d_STdC/README.txt]: So, your preferred option would be:
1. Package w/o examples
2. Remove all README files under drivers
3. Keep the top README file w/o reference to example

Correct?
I'm ok with it. The only thing is that everytime I have to re-align the package there is an
extra effort (but not that much) to clean the package up.

Comment (erwango) [reply ext/hal/st/stmemsc/a3g4250d_STdC/README.txt]: This is indeed indeed my preferred way to do it.
For the extra pain to do it, I suggest to spend few hours on a script to do it for  you.
I have this for cube packages update and this saves me painful hours each time.

**Comment by avisconti** [comment]
rebased and repushed as agreed with @erwango

**Comment by avisconti** [comment]
@erwango 
It seems that you are in charge of everything under ext/hal/st.
This will include also stmemsc.

**Comment by avisconti** [comment]
I got now that I have to solve a couple of false postives in UNDEF_KCONFIG_WHITELIST.
:-(

**Comment by erwango** [comment]
> @erwango
It seems that you are in charge of everything under ext/hal/st.
This will include also stmemsc.

@avisconti, as discussed, can you add your name in front of stmemsc in CODEOWNER?

**Comment by avisconti** [comment]
> @avisconti, as discussed, can you add your name in front of stmemsc in CODEOWNER?

Yes I did.
It is in 52bc149 commit. Is it not matching your expectation? If I remember well we concluded
to have the 'backup' line (i.e. /ext/hal/st) first and then the line for stmemsc, correct?

Now I made a change in the ci-tools whitelist in order to have the CI passing for this PR.

**Comment by erwango** [comment]
> It is in 52bc149 commit. Is it not matching your expectation?

It's perfect, much more than my monday eyes.

File: ext/hal/st/stmemsc/README
Comment (erwango) [reply ext/hal/st/stmemsc/README]: @avisconti , I'm updating the link for Cube packages: https://github.com/zephyrproject-rtos/zephyr/pull/15731

Comment (avisconti) [reply ext/hal/st/stmemsc/README]: > @avisconti , I'm updating the link for Cube packages: #15731

OK, I'm going to change it as well. Thx!

**Comment by avisconti** [comment]
recheck

**Comment by avisconti** [comment]
@erwango  @galak 
I see that now the CI is giving an issue when running tests. The issue is that to properly compile we
need to define following macro:

CONFIG_NEWLIB_LIBC=y
CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y

Before I used 'select' directly in stmemsc Kconfig directly, but I saw that usage of select for visible
symbols (the ones with a prompt) is not recommended. So, I recently changed it in this PR avoiding
'select' and using instead 'depends on'. But this requires the symbols to be defined somewhere during
CI test. I have now two options:

1. Going back to using 'select'
2. Defining somewhere the two symbols. Maybe in sensors_i_z.conf (and all other sensors conf files)
    but this would impact other sensors too. 

So, can you help me in using the right approach for fixing CI?

**Comment by galak** [comment]
> @erwango @galak
> I see that now the CI is giving an issue when running tests. The issue is that to properly compile we
> need to define following macros:
> 
> CONFIG_NEWLIB_LIBC=y
> CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
> 
> Before I used 'select' directly in stmemsc Kconfig file, but I saw that usage of 'select' for visible
> symbols (the ones with a prompt) is not recommended. So, I recently changed it in this PR avoiding
> 'select' and using instead 'depends on'. But this requires the symbols to be defined somewhere during
> CI test. I have now two options:
> 
> 1. Going back to using 'select'
> 2. Defining somewhere the two symbols. Maybe in sensors_i_z.conf (and all other sensors conf files)
>    but this would impact other sensors too.
> 
> So, can you help me in using the right approach for fixing CI?
> 
> EDIT:
> Maybe also @MaureenHelm may add some comments.
> 
> EDIT2:
> It seems that there is a third options discussed with @MaureenHelm, which is to add a new test
> for sensor using the new stmemsc external hal, where I can define a separate conf file.

So I think you should have a separate test for this if newlib is required for the sensor drivers.  We shouldn't burden sensors_i_z.conf with that requirement.

If that makes its easier to use 'depends' than great.  However, I'd try 'imply' and see how that works.

**Comment by avisconti** [comment]
re-pushed again
Now I'm using 'imply' inside the stmemsc Kconfig.
Moreover I defined two new sensor tests for LIS2DW12: test_build_sensors_stmemsc and test_build_sensors_stmemsc_trigger.
All ST sensors that will make use of stmemsc hal in the future will be put inside this test.

**Comment by avisconti** [comment]
@erwango @MaureenHelm 
Not sure what the Codeowners failed check is about, the details are not clear enough.

I think the code now is in a good shape, isn't it?

**Comment by erwango** [comment]
@avisconti , seems there was an issue in CODEOWNER file that prevented correct parsing.
It is fixed now. Can you rebase to take fix into account and see if fixing you issue?

**Comment by avisconti** [comment]
> @avisconti , seems there was an issue in CODEOWNER file that prevented correct parsing.
> It is fixed now. Can you rebase to take fix into account and see if fixing you issue?

OK, thx!
re-pushed! Let's see how it goes now.

**Review by erwango — APPROVED** [review]

**Comment by avisconti** [comment]
@MaureenHelm 
What's ur opinion about introducing the STdC HAL i/f?
If you approve I can later go further porting also LPS22HH and LSM6DSO on it.
Thx!

**Comment by avisconti** [comment]
@galak @MaureenHelm 
Anything else recommended in this PR?

**Comment by avisconti** [comment]
@erwango 
Can you confirm again the license (BSD-3-Clause) is not an issue?
I think it is the same as STM32cube, right?

**Comment by erwango** [comment]
@avisconti , BSD-3 Clause is fine, sure. 
Though, there is a indeed a shade vs Apache which is too subtle for the License bot to handle. This is hence up to the maintainer to ack.

**Comment by erwango** [comment]
@nashif, do you want to merge this one now? or latter once the modules are there?

**Comment by avisconti** [comment]
@nashif @erwango 
I rebased this PR on the latest code.
Moreover I changed the ext/hal/st/stmemsc/Kconfig file with two simple changes:
- I added the USE symbols for all drivers (before was only lis2dw12)
- I removed the prompt from the USE symbols, as we want to force them from the KConfig of each specific driver.

**Comment by avisconti** [comment]
@nashif @erwango 
I moved the STdC in hal_st module. I'm going to close this PR.
The lis2dw12 porting on top of STdC HAL i/f will be resubmitted as a new PR later on.

**Pull request closed without merging**
